mlabs/src/Mlabs/NFT/Validation.hs

```diff
@@ -70,284 +63,297 @@ import Ledger.Typed.Scripts (
-    newPrice1 == newPrice2
-      && cs1 == cs2
-  _ == _ = False
-
-asRedeemer :: UserAct -> Redeemer
+
+import Data.Maybe (catMaybes)
+import Mlabs.NFT.Types (
+  DatumNft (..),
+  InformationNft (
+    info'author,
+    info'id,
+    info'owner,
+    info'price,
+    info'share
+  ),
+  MintAct (Initialise, Mint),
+  NftAppInstance (appInstance'Address, appInstance'AppAssetClass),
+  NftAppSymbol (app'symbol),
+  NftId (nftId'contentHash),
+  NftListHead (head'appInstance),
+  NftListNode (node'appInstance, node'information, node'next),
+  Pointer (pointer'assetClass),
+  UserAct (..),
+  UserId (getUserId),
+  getAppInstance,
+  getDatumPointer,
+  nftTokenName,
+ )
+
+asRedeemer :: PlutusTx.ToData a => a -> Redeemer
 asRedeemer = Redeemer . PlutusTx.toBuiltinData
 
 {-# INLINEABLE mkMintPolicy #-}
 
 -- | Minting policy for NFTs.
-mkMintPolicy :: Address -> TxOutRef -> NftId -> () -> ScriptContext -> Bool
-mkMintPolicy !stateAddr !oref (NftId _ token !outRef) _ !ctx =
-  -- ? maybe author could be checked also, their key should be in signatures.
-  traceIfFalse "UTXO not consumed" hasUtxo
-    && traceIfFalse "Wrong amount minted" checkMintedAmount
-    && traceIfFalse "Does not pay to state" paysToState
-    && traceIfFalse "NFTid TxOutRef and minting TxOutRef are different" sameORef
+mkMintPolicy :: NftAppInstance -> MintAct -> ScriptContext -> Bool
+mkMintPolicy appInstance act ctx =
+  case act of
+    Mint nftid ->
+      traceIfFalse "Only pointer of first node can change." firstChangedOnlyPtr
+        && traceIfFalse "Only One Token Can be Minted" (checkMintedAmount nftid)
+        && traceIfFalse "Old first node must point to second node." (first `pointsTo'` second)
+        && traceIfFalse "New first node must point to new node." (newFirst `pointsTo` newInserted)
+        && traceIfFalse "New node must point to second node." (newInserted `pointsTo'` second)
+        && traceIfFalse "New price cannot be negative." priceNotNegative'
+        && traceIfFalse "Currency symbol must match app instance" checkCurrencySymbol
+        && traceIfFalse "Minted token must be sent to script address" (checkSentAddress nftid)
+        && traceIfFalse "Nodes must be sent to script address" checkNodesAddresses
+    Initialise ->
+      traceIfFalse "The token is not present." True -- todo
+        && traceIfFalse "Only One Unique Token Can be Minted" True -- todo
+        && traceIfFalse "The token is not sent to the right address" True -- todo
   where
-    !info = scriptContextTxInfo ctx
-
-    !hasUtxo =
-      any (\inp -> txInInfoOutRef inp == oref) $
-        txInfoInputs info
-
-    !checkMintedAmount = case flattenValue (txInfoMint info) of
-      [(cur, tn, !val)] ->
-        ownCurrencySymbol ctx == cur
-          && token == tn
-          && val == 1
-      _ -> False
+    ------------------------------------------------------------------------------
+    -- Helpers
 
-    !paysToState = any hasNftToken $ txInfoOutputs info
```

Comment:
here and everywhere we change the bang-pattern'ed bindings for ones without -- please add BangPatterns everywhere back in